### PR TITLE
graalvmXX-ce: new update script & init on aarch64-darwin

### PR DESCRIPTION
**Pull request by tricktron**
###### Description of changes

@bennyandresen @volt  @hlolli @glittershark @babariviere @ericdallo @thiagokokada

Adds two new things:
###### A new update script written in nix

I always wanted to write an update script in nix and chose this derivation. The new script has the following advantages:
- It allows to specify the sources declaratively in the default.nix:
```nix
graalvm17-ce-dev = mkGraal rec {
    config = {
      aarch64-darwin = {
        arch = "darwin-aarch64";
        products = [
          "graalvm-ce"
          "native-image-installable-svm"
        ];
      };
    };
    version = "22.2.0-dev-20220330_2302";
    javaVersion = "17";
    platforms = builtins.attrNames config;
  };
```

This then helps to write tests in the derivation by checking for the supported products:
`lib.optionalString (builtins.any (a: a == "python-installable-svm") platform.products) ''`

- Generates sources per derivation. Updating the sources for `graalvm17-ce` does not update the sources of `graalvm11-ce`.
- Downloads the `.sha256` file of the release source files instead of downloading the whole `.zip/.tar.gz` files and then computing the sha256. This saves time.

###### aarch64-darwin dev build
`aarch64-darwin` is now supported on the dev builds. The new update script allowed me to easily add it. The plan is to only keep the dev builds around until the aarch64-darwin is released to stable upstream because I don't think it is worth maintaining all the dev-builds. But now we could and this is open for discussion.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/graalvm/community-edition/default.nix [hunk 1]
```diff
@@ -20,8 +46,33 @@ in
...
+        products = commonProducts ++ [ "python-installable-svm" ];
+      };
+    };
     version = "22.0.0.2";
     javaVersion = "17";
-    platforms = [ "x86_64-linux" "x86_64-darwin" ];
+    platforms = builtins.attrNames config;
+  };
+
+  graalvm17-ce-dev = mkGraal rec {
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/default.nix:64]: What is the idea when `graalvm17` final is out to `darwin-aarch64`, to alias `graalvm-17-ce-dev = `graalvm17-ce`?

Ideally I would prefer that we don't add a new package and if possible it would be better to have `graalvm17-ce` for  `darwin-aarch64` to map to `dev` versions temporary. But if this is not possible I am fine with the solution above too.

Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:64]: Aliasing would work.

If we decide to drop the env builds we just remove the `graalvm17-ce-dev` derivation again and add the `aarch64-darwin` platform to:
```nix
graalvm17-ce = mkGraal rec {
    config = {
      x86_64-darwin = {
        arch = "darwin-amd64";
        products = commonProducts ++ [ "python-installable-svm" ];
      };
      x86_64-linux = {
        arch = "linux-amd64";
        products = commonProducts ++ [ "python-installable-svm" ];
      };
      aarch64-darwin = {
        arch = "darwin-aarch64";
        products = [ "supported products on darwin-aarch64 on release" ];
      };
    };
    version = "22.0.0.2";
    javaVersion = "17";
    platforms = builtins.attrNames config;
  };
```

Comment (thiagokokada) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:64]: We can never remove a name though, this is why aliasing would be necessary in this case.

Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:64]: @thiagokokada You are right. We would have to add something like this to `pkgs/top-level/aliases.nix`:
```nix
graalvm17-ce-dev = throw "graalvm17 dev builds were discontinued in ..."
```
This is a bit ugly.

The next graalvm release will be on April 26, 2022: 22.1.0, 21.3.2. Maybe it makes sense 
> to have graalvm17-ce for darwin-aarch64 to map to dev versions temporary.

until then. I'll have a look at it tomorrow.

**Comment by thiagokokada** [comment]
Another thing: `buildGraalvmNativeImage` defaults to `graalvm11-ce`, and AFAIK no packages overrides it. This mean this PR is not sufficient to build GraalVM programs on `aarch64-linux`.

We can probably override this default just to `aarch64-linux` for now, of upstream doesn't plan to release a version of GraalVM 11 for aarch64.

**Review by thiagokokada — COMMENTED** [review]
In general this is a great PR. Adding some comments though, and not approving it yet because of the `graalvm17-ce-dev` thing.

But yeah, I really like what you did here. Thank you!

File: pkgs/development/compilers/graalvm/community-edition/update.nix [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        jq . ${sourcesJson} > ${sourcesJsonPath}
+        drvName=$(awk '/${name}/{ print NR; exit }' ${defaultNixPath})
+        awk -v drvName="$drvName" -i inplace \
+        'NR>drvName {sub(/${graalVersion}/, "${newVersion}")} 1' ${defaultNixPath}
+      ''
+    else ''echo "No new version found. Skip updating."'';
+
+in
+writeScript "update-graal.sh" ''
+  #!${runtimeShell}
```
[the remark this answers is not in the document]
Comment (thiagokokada) [reply pkgs/development/compilers/graalvm/community-edition/update.nix:211]: Actually it should be:

```suggestion
writeShellScript "update-graal.sh" ''
```

[the remark this answers is not in the document]
Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/update.nix:211]: You are absolutely right👍.

File: pkgs/development/compilers/graalvm/community-edition/mkGraal.nix [hunk 3]
```diff
@@ -86,7 +81,7 @@ let
       zlib
     ];
 
-    nativeBuildInputs = [ unzip perl autoPatchelfHook makeWrapper ];
+    nativeBuildInputs = [ unzip perl makeWrapper ] ++ lib.optional stdenv.isLinux [ autoPatchelfHook ];
```
[the remark this answers is not in the document]
Comment (thiagokokada) [reply pkgs/development/compilers/graalvm/community-edition/mkGraal.nix:84]: You can either try to match the exactly change in `staging` (this may or may not work), or rebase your PR based on `staging` (this is safer however it also means your PR will only go to `master` on the next `staging` merge).

File: pkgs/development/compilers/graalvm/community-edition/update.nix [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+  ...
+  },
+  "platform/arch2": {
+  ...
+  }
+  }
+*/
+
+let
+  productJavaVersionGraalVersionSep = "|";
```
[the remark this answers is not in the document]
Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/update.nix]: Good point, I named it just `separator`.

File: pkgs/development/compilers/graalvm/community-edition/default.nix [hunk 5]
```diff
@@ -1,17 +1,62 @@
...
 {
   inherit mkGraal;
 
   graalvm11-ce = mkGraal rec {
-    version = "22.0.0.2";
+    config = {
+      x86_64-darwin = {
+        arch = "darwin-amd64";
+        products = commonProducts ++ [ "python-installable-svm" ];
+        version = graalvm11-ce-release-version;
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/default.nix:37]: Maybe we can pass this as a parameter on `mkGraal`, and if someone overrides it in the config we them use this version.

This way, we don't need to explicitly pass the version for every other platform (just `aarch64-linux`).

Comment (thiagokokada) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:37]: Doing it this way we wouldn't even need those constants. We can just use the version directly because there wouldn't be repetitions.

Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/default.nix]: I am not sure if I understand you completely. I have extracted the commonConfig to a `baseConfig` to reduce the repetition in the latest commit. If you meant something else, could you please elaborate?

Comment (thiagokokada) [reply pkgs/development/compilers/graalvm/community-edition/default.nix]: I was thinking more of something like this. On `mkGraal.nix`, you can add one more parameter:

```nix
{ javaVersion
, platforms
, config
, defaultVersion
, useMusl ? false
}:
```

Then you could do something like this:

```nix
graalvm17-ce = mkGraal rec {
  defaultVersion = "22.0.0.2";
  config = {
    aarch64-darwin = {
      arch = "darwin-aarch64";
      products = [
        "graalvm-ce"
        "native-image-installable-svm"
      ];
      version = "22.2.0-dev-20220331_1955";
    };
  };
};
```

This way there is no repetition, and it is very clear what is happening IMO.

File: pkgs/development/compilers/graalvm/community-edition/mkGraal.nix [hunk 6]
```diff
@@ -33,16 +32,19 @@
...
-  platform = {
-    aarch64-linux = "linux-aarch64";
-    x86_64-linux = "linux-amd64";
-    x86_64-darwin = "darwin-amd64";
-  }.${stdenv.system} or (throw "Unsupported system: ${stdenv.system}");
+  platform = config.${stdenv.hostPlatform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
+  version = platform.version;
+  name = "graalvm${javaVersion}-ce";
+  sourcesFilename = "${name}-sources.json";
+  sources = builtins.fromJSON (builtins.readFile (./. + "/${sourcesFilename}"));
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/mkGraal.nix:47]: This should also work:
```suggestion
  sources = builtins.fromJSON (builtins.readFile (./${sourcesFilename}));
```

File: pkgs/development/compilers/graalvm/community-edition/mkGraal.nix [hunk 7]
```diff
@@ -194,9 +164,15 @@ let
...
         # Set JAVA_HOME automatically.
         mkdir -p $out/nix-support
-        cat <<EOF > $out/nix-support/setup-hook
+        cat > $out/nix-support/setup-hook << EOF
           if [ -z "\''${JAVA_HOME-}" ]; then export JAVA_HOME=$out; fi
         EOF
+        ${
+        lib.optionalString (stdenv.isLinux) ''
+          ${copyClibrariesToOut "$out/lib/svm/clibraries"}
+          ${copyClibrariesToLib}
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/mkGraal.nix:173]: I don't think those two functions make much sense anymore. This was when we used to apply a specific configuration for GraalVM 8 vs GraalVM 11.

So I think we can just add the text directly here. WDYT?

File: pkgs/development/compilers/graalvm/community-edition/default.nix (hunk 5 above)
Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/default.nix]: That is what we had before and I tried that but this leads to a problem in the update script when creating the
`product|javaVersion|graalVersion` for each platform.

Since we have now two possible versions, dev and release depending on the platform, I need to specify them per platform. Otherwise the defaultVersion overwrites the dev version in the `graalvmXX-ce-sources.json` file or vice versa.

That was one of the reasons why I first added a new `graalvmXX-ce-dev` name because then I had only one version per derivation, eg. dev for `graalvmXX-ce-dev` and release for `graalvmXX-ce` so I didn't have to distinguish.

I don't like the current repetition either but we need it temporarily. Since we are planning to remove the dev versions when `aarch64-darwin` makes it into the release, we can then remove this repetition too.

File: pkgs/development/compilers/graalvm/community-edition/mkGraal.nix (hunk 6 above)
Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/mkGraal.nix:47]: Thanks, I'll change it.

File: pkgs/development/compilers/graalvm/community-edition/default.nix (hunk 5 above)
Comment (thiagokokada) [reply pkgs/development/compilers/graalvm/community-edition/default.nix]: Maybe I am missing something, but not sure why the proposed solution wouldn't work. You can use `config.version` if it exists, or use `defaultVersion` if it doesn't (of course, you need pass `defaultVersion` to the update script as an argument too). Something like this:

```
nix-repl> {}.version or "22.0.0.5"
"22.0.0.5"
```

Where `{}` would be the config attrset.

Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/default.nix]: Ah now I get it:smiley:. Sure that works. Let me quickly implement that.

File: pkgs/development/compilers/graalvm/community-edition/mkGraal.nix (hunk 7 above)
Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/mkGraal.nix:173]: Jep.

File: pkgs/development/compilers/graalvm/community-edition/default.nix (hunk 5 above)
Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/default.nix]: Ok, I have added the `defaultVersion` but went back to a config block for `graalvm11-ce` and `graalvm17-ce` because I like to explicitly configure it there instead of creating a baseConfig set.

**Comment by tricktron** [comment]
@thiagokokada at some point, we should also check that this builds on `aarch64-linux` because if I run it through my remote docker builder it fails during the jshell test. However, it passes the native image tests so maybe we can fix the following todo:
https://github.com/NixOS/nixpkgs/blob/2d7d3692c125cc07954b342f84c797019c37633c/pkgs/development/compilers/graalvm/community-edition/default.nix#L60-L64

File: pkgs/development/compilers/graalvm/community-edition/mkGraal.nix (hunk 6 above)
Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/mkGraal.nix]: Looks like this change fails the output comparison test. Is it ok if I revert?

**Comment by thiagokokada** [comment]
> @thiagokokada at some point, we should also check that this builds on `aarch64-linux` because if I run it through my remote docker builder it fails during the jshell test. However, it passes the native image tests so maybe we can fix the following todo:
> https://github.com/NixOS/nixpkgs/blob/2d7d3692c125cc07954b342f84c797019c37633c/pkgs/development/compilers/graalvm/community-edition/default.nix#L60-L64
> 
> Can you test this on `aarch64-linux`?
> 

I have no access to an `aarch64-linux` machine with the exception of running this through emulation (that is very slow).

You can always use of org too.

File: pkgs/development/compilers/graalvm/community-edition/mkGraal.nix (hunk 6 above)
Comment (thiagokokada) [reply pkgs/development/compilers/graalvm/community-edition/mkGraal.nix]: Sure.

File: pkgs/development/compilers/graalvm/community-edition/default.nix [hunk 8]
```diff
@@ -1,17 +1,60 @@
...
   };
+
+  /*
+    Looks a bit ugly but makes version update in the update script using sed
+    much easier
+  */
+  graalvm11-ce-release-version = "22.0.0.2";
+  graalvm17-ce-release-version = "22.0.0.2";
+  graalvm11-ce-dev-version = "22.2.0-dev-20220401_1942";
+  graalvm17-ce-dev-version = "22.2.0-dev-20220401_1942";
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/default.nix:19]: Minor nitpick but I think we can just pass this values directly to the `mkGraal` function now. There is no more unnecessary repetition and I think it makes easier to understand which version is being used where.

**Review by thiagokokada — COMMENTED** [review]
Well, it is looking good now. Not sure why the eval is failing though, let me try again (but it may be an issue somewhere in your code too).

@ofborg eval
@ofborg build babashka

**Comment by thiagokokada** [comment]
BTW, all my comments mentioning setting `graalvm17-ce-dev` as default for `aarch64-linux`: please substitute to `aarch64-darwin`.

File: pkgs/development/compilers/graalvm/community-edition/default.nix (hunk 8 above)
Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:19]: Yes, but then the version replacement in the update script gets much more complex:

E.g I then need to use some awk or sed magic to find the graalvm11-ce line and replace the defaultVersion or version string exactly once and the same if the update script is invoked with graalvm17.

Alternatively, I could probably also do that version update using nix. I need to think about this.

Comment (thiagokokada) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:19]: Ah ok, this seems fair.

**Comment by thiagokokada** [comment]
@ofborg build babashka clojure-lsp HentaiAtHome

**Comment by tricktron** [comment]
@ofborg build graalvm11-ce

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
I think this is a good PR to start. `aarch64-darwin` is failing for `babashka`, but we can always fix this on `master` (since it didn't even try to build before).

**Comment by thiagokokada** [comment]
Thanks @tricktron !

**Comment by tricktron** [comment]
> Thanks @tricktron !

Thank you for the quick and helpful reviews😃.
